src/System.Private.Uri/src/System/PooledCharArray.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+
+namespace System
+{
+    internal struct PooledCharArray
+    {
+    	private char[] _buffer;
+
+    	public PooledCharArray(int size)
+    	{
+    		_buffer = ArrayPool<char>.Shared.Rent(size);
+    	}
+
+		public PooledCharArray(char[] fromArray)
+		{
+			_buffer = ArrayPool<char>.Shared.Rent(fromArray.Length);
+			for (int i = 0; i < fromArray.Length; i++)
+			{
+				_buffer[i] = fromArray[i];
+			}
+		}
+
+    	public void Release()
+    	{
+    		if (_buffer != null)
+    		{
+    			ArrayPool<char>.Shared.Return(_buffer);
+    			_buffer = null;
+    		}
+    	}
+
+    	public int Length => _buffer.Length;
+
+		public char[] Array => _buffer;
+
+    	public string GetStringAndRelease(int stringLength)
+    	{
+    		string ret = new string(_buffer, 0, stringLength);
+    		Release();
+    		return ret;
+    	}
+
+		public void CopyAndRelease(char[] dest, int offset, int count)
+		{
+			while (count > 0)
+			{
+				dest[offset] = _buffer[offset++];
+				count--;
+			}
+			Release();
+		}
+
+    	public char this[int index]
+    	{
+    		get
+    		{
+    			return _buffer[index];
+    		}
+    		set
+    		{
+    			_buffer[index] = value;
+    		}
+    	}
+
+        public unsafe bool IsSameString(char* ptrStr, int start, int end)
```

Comment:
also, you can eliminate your Release() method.

it's better to force them to release at the same time they "get what they want" as that way it's hard to forget to Release()

---
In reply to: [103758643](https://github.com/dotnet/corefx/pull/16456#discussion_r103758643) [](ancestors = 103758643)